File: Graph/841. Keys and Rooms/841.java
class Solution {
    public void go(int room_num, List<List<Integer>> rooms, boolean[] visited){
        visited[room_num] = true;
        for(int key : rooms.get(room_num)){
            if(!visited[key]){
                go(key, rooms, visited);
            }
        }
    }

    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];
        go(0, rooms, visited);

        for(boolean status : visited){
            if(!status){return false;}
        }

        return true;
    }
}

File: 13-Graph/DFS/841. Keys and Rooms/841.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = set()

        def go(room_num):
            visited.add(room_num)
            for key in rooms[room_num]:
                if key not in visited:
                    go(key)

        go(0)

        return len(visited) == len(rooms)
